clash-prelude/src/Clash/Class/Resize.hs

```diff
@@ -117,3 +120,57 @@ checkedTruncateB ::
   f (a + b) -> f a
 checkedTruncateB v =
   checkIntegral (Proxy @(f a)) v `seq` truncateB v
+
+-- | Helper function of 'maybeFromIntegral', 'maybeResize' and 'maybeTruncateB'
+maybeIntegral ::
+  forall a b.
+  (Integral a, Integral b, Bounded b) =>
+  Proxy b ->
+  a -> Bool
+maybeIntegral Proxy v =
+     toInteger v < toInteger (maxBound @b)
+  && toInteger v > toInteger (minBound @b)
+
+-- | Like 'fromIntegral', but returns 'Nothing' if /a/ is out of bounds for /b/.
+-- Useful when you do not know /a/ can be out of bounds, and would like to have
+-- your assumptions checked.
+
+-- * __NB__: 'fromIntegral' is not well suited for Clash as it will go through
+-- 'Integer' which is arbitrarily bounded in HDL. Instead use
+-- 'Clash.Class.BitPack.bitCoerce' and the 'Resize' class.
+maybeFromIntegral ::
+  forall a b.
+  (Integral a, Integral b, Bounded b) =>
+  a -> Maybe b
+maybeFromIntegral v =
+  if maybeIntegral (Proxy @b) v
+    then Just (fromIntegral v)
+    else Nothing
+
+-- | Like 'resize', but returns 'Nothing' if /f a/ is out of bounds for /f b/.
+-- Useful when you do not know /f a/ can be out of bounds, and would like to
+-- have your assumptions checked.
+maybeResize ::
```

Comment:
This will break for anything larger than native word size (64 bit). At least mention this limitation in the doc.

Reply:
Oh in fact you can do a short circuit on the type level if `BitSize a >= BitSize b` which means this function will be free in that case in hardware.